### ISSUE TITLE
[BUG] Select a USB drive as an install target

### ISSUE THREAD

**Issue by polarathene** [calamares/calamares#879]
#### Info regarding which version of Calamares is used, which Distribution
Latest in Manjaro 17.1-RC1 KDE

#### Provide information on how the disks are set up, in detail, with full logs of commands issued 
Used Etcher to flash the ISO to a USB drive, had another 32GB USB drive(formatted as far as I know, previously had a manjaro ISO flashed to it, later failed install attempt via manjaro-architect via the manjaro 17.1RC live media(live media often becomes unstable/buggy if it's been running too long).

#### What do you expect to have happen when Calamares installs?
Be able to select the 32GB USB drive as a target for installing to instead of only the internal HDD.

#### Describe the issue you encountered
In the various states for the target USB described above, Calamares failed to list the drive as an installable target. The install USB was also not listed(although I imagine that's intentional). The internal HDD of the laptop was detecte with MBR(USB target was formatted with MBR).

#### Steps to reproduce the problem
- Format a USB with MBR partiton table
- Boot the Manjaro-17.1RC KDE on an x86_64 BIOS(not UEFI) system(ATI drivers, Intel Core 2 Duo dual core system, USB 2.0). I booted this via another USB /dev/sdb if that helps.
- Run Calamares installer, progress to storage selection for where to install.
- Note that the blank USB drive(/dev/sdc) does not appear as a valid target.
- manjaro-architect detects /dev/sdc just fine.

**Comment by abucodonosor** [comment]
Where is the installation log ?

**Comment by polarathene** [comment]
@abucodonosor I filed this later in the day, I've managed to get the system installed via manjaro-architect ISO which had no problem detecting the USB drive. I'll try find some time to flash the mentioned media above and reproduce(probably with another USB as this 32GB is currently in use). I'll provide a log when I've done so.

Anything in particular from the log you're after? Just the issue of Calamares not listing the USB device for install target correct? I don't see any point in reproducing the instability over time running live media, that's common on many distro live media on various systems in my experience.

**Comment by abucodonosor** [comment]
We exclude all devices containing iso9660 ..

```
had another 32GB USB drive(formatted as far as I know, previously had a manjaro ISO flashed ...)

```
just normal formating won't help you here , you need do something like this:

dd if=/dev/zero of=/dev/sdX bs=1k count=2048 && sync
parted /dev/sdX mklabel msdos

**Comment by polarathene** [comment]
Is there any reason to do that? Calamares provides the tooling to format a drive, so what is wrong with wanting to have Calamares detect the USB and format that drive to install the OS on? Does iso9660 cause some sort of problem?

**Comment by abucodonosor** [comment]
@polarathene 

Yes it will since your USB you boot the Live image is iso9660 ..
Meaning we would allow to install into the driver you boot the system
and that isn't going to work :)

**Comment by polarathene** [comment]
Well I understand that, but I'm not going to do that. If you are concerned about that could you not check the media for some file that would be on the live image? Or some better way to isolate the install media?

Alternatively just add some warning symbol/icon next to it or a popup confirmation dialog, still allowing a user to select the storage device as a target but be warned that it might be the live image of the installer and to take caution?

Have many users raised this as an issue that being able to select the storage device the installer is loaded from has been accidentally selected as the install target and caused issues?

**Comment by abucodonosor** [comment]
There is no sane way to check that on USB devices, iso9660 on a CD/DVD/BR etc
are marked at least read-only but is not the case on an USB device with iso9660.

Is kind allow all or forbid all..

Not so many but it was added because some did that .. boot from /dev/sdX USB
and want to install in the same device..

**Comment by philmmanjaro** [comment]
We have several commits in this regard:

* Exclude ISO9660 volumes https://github.com/calamares/calamares/commit/6c928e04040b4d7d5373545a94ff81df33b67774
* Be more explicit on why a device is winnowed from the list https://github.com/calamares/calamares/commit/360a114ed43678695f7ab829cd00c1db7e8e9cc1

**Comment by polarathene** [comment]
What about some button on the UI to refresh/allow iso9660 devices, could have a tooltip or popup dialog warning about the fact that if the live image is booted from a USB device it will be listed as an available install target as a result, and to take caution not to choose it?

**Comment by philmmanjaro** [comment]
@polarathene: install medias will be normally written via **dd** and therefore they are not usable. All the space is used for the ISO. When you reformat it and remove any ISO9660 tags it should be available. Give me a sane reason why we shouldn't block ISO9660 as I see any.

**Comment by adriaandegroot** [comment]
There's a fistful of heuristics in play to determine what devices make sense to install to. They may be overly agressive -- or maybe not. What you want to avoid is accidentally installing to the same USB stick you booted from -- I think I can confidently state that we're never going to support booting from the first half of a 32GB stick and installing to the other half of the same stick.

Calamares excludes devices (whole devices) which:
 - is too small for the target image
 - contain the mounted /
 - is a zram device (compressed ramdisk, since that doesn't make any sense)
 - contain any mounted partition (there's discussion about this, since some systems are very aggressive about automounting)
 - has any partition for which blkid reports iso9660 (that's what this issue is about .. there's another issue which also mentions this)

The debug message when winnowing iso9660 is that it's probably a CD -- but I expect that you want to exclude **read-only** iso9660s to identify those. Unless CD-RW makes the read-only determination a mess. The only other case I can think of, is that a device shows up more than once in the list obtained from KPMCore, and removing the / device leaves the actual-physical-device-with-the-live-image-on-it in the list.

So, yeah, I'm wondering now why we drop iso9660 devices, instead of testing in a better way.

**Comment by adriaandegroot** [comment]
#545 is the other issue ..

**Comment by polarathene** [comment]
@philmmanjaro  > Give me a sane reason why we shouldn't block ISO9660 as I see any.

A reason... well this issue is one. I want to install to another USB. Manjaro and Calamares is about catering to user friendliness yes? I created this issue out of confusion why my device was blacklisted/missing from the storage options while other installers do not.

You're blacklisting it because of the potential that it could be the install media and users can accidentally target it. It's a UX issue. Rather than allow a user to still choose the media and indicate/warn them that it may be install media and to take caution proceeding with it as an install target, you remove the option completely, preventing the user from using a valid target.

In this case it's due to the target media most likely having something like a live image on it. So yes, now knowing why it's been excluded from a selection, I could go and format it, or Calamares could do this for me after I confirm I'm ok with it. That'd be a UX friendly thing to do, let me select the USB, warn me, then prompt to format if required so that it can be used for an install target, I don't have to side track from Calamares, no confusion arises.

Getting a USB to flash install media onto would be a pain if they all refused to flash to ISO9660 media. Quite a few options usually warn you that the media will lose all it's data, others just expect you know what you're doing. If they refused, that'd be a tad annoying and perhaps confusing to some users without knowing why their media was refused or unavailable to selection when as far as they know that shouldn't be the case.

**Solution:** UX, add a triangle with exclamation mark warning symbol next to the storage media(either during selection in combobox or reveal the symbol when selection is made if appropriate). When user proceeds to next step, interrupt with a dialog confirmation/warning on risk of writing to/formatting the media(eg. it could be the install media).

Now the users who made this mistake can avoid it thanks to better UX informing them of the potential mistake, while users who know the install target is fine can actually use it without having to do extra work(flashing install media onto a USB which already has existing install media doesn't require a user to format it prior to do so). You already allow a user to select an internal drive with existing OS like Windows on it for example and format the drive for the installer to have full use of. How is this different?

---

@adriaandegroot > What you want to avoid is accidentally installing to the same USB stick you booted from -- I think I can confidently state that we're never going to support booting from the first half of a 32GB stick and installing to the other half of the same stick.

Understandable. I was not wanting either of those. I had a 16GB USB with install media, and a 32GB USB as an install target. Not wanting to install on the install media is a rather obvious one to want to avoid. If you have no way to identify the media being used to run the live image, that's unfortunate. It sounds like you do from the strategies listed, that the install media should be identifiable before requiring to filter out by iso9660 medias?

**Comment by adriaandegroot** [comment]
Is it possible for devices to end up listed twice? E.g. / is /dev/onewaytothinkaboutdevices/dev0 but at the same time /dev/anotherway/dev0 is the same physical partition / on the same physical device? Because that's the hard case: finding aliases that will accidentally allow you to destroy the install media.

**Comment by jonathonf** [comment]
If Calamares excludes devices with the currently mounted root (e.g. the installer USB device), is there any reason to exclude iso9660 devices too? Given Calamares has disk partitioning/formatting capabilities it's possibly reasonable to expect it to be able to format and prepare a device for installation, and a user won't necessarily know that writing an installer image to a USB drive makes it identify as an iso9660 device; rather, the (otherwise perfectly valid) USB flash drive simply doesn't show as a target.

**Comment by enderger** [comment]
Test for CALAMARIS

**Comment by kkofler** [comment]
This is _Calamar**e**s_. _Calamar**i**s_ is a completely different project!

**Comment by enderger** [comment]
A discrepancy is a discrepancy

**Comment by zefram-cochrane** [comment]
I just struck something I suspect is this same problem - but more insidious.

I just picked up a neat 1Tb USB key. Marvelous what they can pack into these things these days. I'm keen to move to Manjaro-KDE 18.1.5 (currently on deepin 15.11 ...) so thought I'd install it onto that neat USB and shift over by easy stages. Soon found out that formatting ext4 takes a very long time so decided to use gparted to carve out some sensible partitions by hand and then replace one with Manjaro.

Fired up a live usb - took me a while to figure out I needed to use etcher to do so and wasted some time with unetbootin and then dd first. But finally there I am in Calamares for the first time and I pick the partition to replace on the sdd device and I'm just about to hit the button when I notice in the summary that sda has been selected for the install, not sdd.

Obviously I backtrack to try to figure out where I went wrong and ... I didn't go wrong. Calamares was about to do my a nice favour and clobber my daily driver before I'm ready. So I stop there and google to find out why and ... well, here I am.

Looks like polarathene managed to get good things to happen via manjaro-architect so I'll try that way tomorrow but ... if I hadn't double-checked before pushing the Install button I'd be a very unhappy vegemite tonight!

I suggest that, until this bug is fixed, the "replace a partition" option in Calamares should be deactivated. Too damned dangerous!
